RecoMTD/TrackExtender/plugins/TrackExtenderWithMTD.cc

```diff
@@ -207,7 +207,87 @@ namespace {
     }
     return tofpid;
   }
+
+  bool getTrajectoryStateClosestToBeamLine(const Trajectory& traj,
+                                           const reco::BeamSpot& bs,
+                                           const Propagator* thePropagator,
+                                           TrajectoryStateClosestToBeamLine& tscbl) {
+    // get the state closest to the beamline
+    TrajectoryStateOnSurface stateForProjectionToBeamLineOnSurface =
+        traj.closestMeasurement(GlobalPoint(bs.x0(), bs.y0(), bs.z0())).updatedState();
+
+    if (!stateForProjectionToBeamLineOnSurface.isValid()) {
+      edm::LogError("CannotPropagateToBeamLine") << "the state on the closest measurement isnot valid. skipping track.";
+      return false;
+    }
+
+    const FreeTrajectoryState& stateForProjectionToBeamLine = *stateForProjectionToBeamLineOnSurface.freeState();
+
+    TSCBLBuilderWithPropagator tscblBuilder(*thePropagator);
+    tscbl = tscblBuilder(stateForProjectionToBeamLine, bs);
+
+    return tscbl.isValid();
+  }
+
+  bool trackPathLength(const Trajectory& traj,
+                       const TrajectoryStateClosestToBeamLine& tscbl,
+                       const Propagator* thePropagator,
+                       double& pathlength) {
+    pathlength = 0.;
+
+    bool validpropagation = true;
+    double pathlength1 = 0.;
+    double pathlength2 = 0.;
+
+    //add pathlength layer by layer
+    for (auto it = traj.measurements().begin(); it != traj.measurements().end() - 1; ++it) {
+      const auto& propresult = thePropagator->propagateWithPath(it->updatedState(), (it + 1)->updatedState().surface());
+      double layerpathlength = std::abs(propresult.second);
+      if (layerpathlength == 0.) {
+        validpropagation = false;
+      }
+      pathlength1 += layerpathlength;
+    }
+
+    //add distance from bs to first measurement
+    if (traj.direction() == alongMomentum) {
+      const auto& propresult2 =
+          thePropagator->propagateWithPath(tscbl.trackStateAtPCA(), traj.firstMeasurement().updatedState().surface());
+      pathlength2 = propresult2.second;
+      if (pathlength2 == 0.) {
+        validpropagation = false;
+      }
+      pathlength = pathlength1 + pathlength2;
+    } else {
+      const auto& propresult2 =
+          thePropagator->propagateWithPath(tscbl.trackStateAtPCA(), traj.lastMeasurement().updatedState().surface());
+      pathlength2 = propresult2.second;
+      if (pathlength2 == 0.) {
+        validpropagation = false;
+      }
+      pathlength = pathlength1 + pathlength2;
```

Comment:
couldn't it be put outside the if ? (deleting line 260 as well)

Reply:
good catch, it can be moved out; the ```if (pathlength2 == 0.) {``` is similarly in common.